server/v2/api/telemetry/server.go

```diff
@@ -1,47 +1,127 @@
 package telemetry
 
 import (
+	"context"
 	"encoding/json"
 	"fmt"
 	"net/http"
 	"strings"
 
-	"github.com/gorilla/mux"
+	"cosmossdk.io/core/transaction"
+	"cosmossdk.io/log"
+	serverv2 "cosmossdk.io/server/v2"
 )
 
-func RegisterMetrics(r mux.Router, cfg Config) (*Metrics, error) {
-	m, err := New(cfg)
+var (
+	_ serverv2.ServerComponent[transaction.Tx] = (*TelemetryServer[transaction.Tx])(nil)
+	_ serverv2.HasConfig                       = (*TelemetryServer[transaction.Tx])(nil)
+)
+
+const ServerName = "telemetry"
+
+type TelemetryServer[T transaction.Tx] struct {
+	config  *Config
+	logger  log.Logger
+	server  *http.Server
+	metrics *Metrics
+}
+
+// New creates a new telemetry server.
+func New[T transaction.Tx]() *TelemetryServer[T] {
+	return &TelemetryServer[T]{}
+}
+
+// Name returns the server name.
+func (s *TelemetryServer[T]) Name() string {
+	return ServerName
+}
+
+func (s *TelemetryServer[T]) Config() any {
+	if s.config == nil || s.config == (&Config{}) {
+		return DefaultConfig()
+	}
+
+	return s.config
+}
+
+// Init implements serverv2.ServerComponent.
+func (s *TelemetryServer[T]) Init(appI serverv2.AppI[T], cfg map[string]any, logger log.Logger) error {
+	serverCfg := s.Config().(*Config)
+	if len(cfg) > 0 {
+		if err := serverv2.UnmarshalSubConfig(cfg, s.Name(), &serverCfg); err != nil {
+			return fmt.Errorf("failed to unmarshal config: %w", err)
+		}
+	}
+	s.config = serverCfg
+	s.logger = logger
+
+	metrics, err := NewMetrics(s.config)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("failed to initialize metrics: %w", err)
 	}
+	s.metrics = metrics
 
-	metricsHandler := func(w http.ResponseWriter, r *http.Request) {
-		format := strings.TrimSpace(r.FormValue("format"))
+	return nil
+}
 
-		gr, err := m.Gather(format)
-		if err != nil {
-			w.Header().Set("Content-Type", "application/json")
-			w.WriteHeader(http.StatusBadRequest)
-			bz, err := json.Marshal(errorResponse{Code: 400, Error: fmt.Sprintf("failed to gather metrics: %s", err)})
-			if err != nil {
-				return
-			}
-			_, _ = w.Write(bz)
+func (s *TelemetryServer[T]) Start(ctx context.Context) error {
+	if !s.config.Enable {
+		return nil
+	}
 
-			return
-		}
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", s.metricsHandler)
+	// keeping /metrics for backwards compatibility
+	mux.HandleFunc("/metrics", func(w http.ResponseWriter, r *http.Request) {
+		http.Redirect(w, r, "/", http.StatusMovedPermanently)
+	})
 
-		w.Header().Set("Content-Type", gr.ContentType)
-		_, _ = w.Write(gr.Metrics)
+	s.server = &http.Server{
+		Addr:    s.config.Address,
+		Handler: mux,
 	}
 
-	r.HandleFunc("/metrics", metricsHandler).Methods("GET")
+	go func() {
+		s.logger.Info("Starting telemetry server", "address", s.config.Address)
+		if err := s.server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			s.logger.Error("Failed to start telemetry server", "error", err)
+		}
+	}()
 
-	return m, nil
+	return nil
```

Comment:
This is not blocking. A http server startup failure is logged but would not be visible otherwise, I assume

Reply:
my bad, it should have been, updated, the whole server components is already in a go routine.